### ISSUE TITLE
RFC: Invenio new uploader

### ISSUE THREAD

**Issue by egabancho** [jirikuncar/invenio#278]
# Idea

Replace the current BibUpload.
# Proposal

Create a new uploader keeping this list of ideas about it in mind:
- Work with JSON natively rather than Marc, although the last could be uploaded as well.
- Run tasks in parallel when possible, for example one huge insert coming from one file could generate several insert tasks that could run in parallel.
- Easy to customise depending on the installation needs. For that we thought about creating _lightweight workflows_ (https://github.com/romanchyla/workflow) for each of the common/current modes, so each installation could override the current ones or implement their owns.
- This lightweight workflows should be made out of small monolithic tasks and it would be nice to have a rollback function for each of this functions in case of error in any of them while running the workflow (reverse workflow?)
- Plug the new documents module with a default document storage (same behaviour as BibDocFile).
- Fire signals before and after running any workflow would be a good thing to "connect" the new uploader with other Invenio tools like `bibindex`, `webcoll`, etc.

See WIP implementation at #135
